extensions/amp-form/0.1/form-verifiers.js

```diff
@@ -0,0 +1,326 @@
+      if (element && element.checkValidity()) {
+        element.setCustomValidity(error.message);
+        errorElements.push(element);
+      }
+    }
+
+    // Remove the dirty flag from the successful groups that have values.
+    for (let i = 0; i < this.groups_.length; i++) {
+      const group = this.groups_[i];
+      if (group.isFilledOut() && group.containsNoElementOf(errorElements)) {
+        group.setDirty(false);
+      }
+    }
+  }
+
+  /**
+   * Check if any group in the form needs to be verified.
+   * @return {boolean}
+   */
+  shouldVerify_() {
+    return this.groups_.some(group => group.shouldVerify());
+  }
+
+  /**
+   * Check if an element is a member of a verification group.
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  isVerificationElement_(element) {
+    return !!this.getGroup_(element);
+  }
+
+  /**
+   * Get the group that contains the given element.
+   * @param {!Element} element
+   * @return {?VerificationGroup}
+   */
+  getGroup_(element) {
+    for (let i = 0; i < this.groups_.length; i++) {
+      const group = this.groups_[i];
+      if (group.containsElement(element)) {
+        return group;
+      }
+    }
+    return null;
+  }
+}
+
+/**
+ * Encapsulates a list of related inputs and a "dirty" flag to
+ * help determine when to send a verification request. Provides methods
+ * for querying the state of the grouped elements.
+ * @private
+ */
+class VerificationGroup {
+  /**
+   * @param {!Array<!Element>} elements
+   */
+  constructor(elements) {
+    /** @private @const */
+    this.elements_ = elements;
+
+    /** @private */
+    this.dirty_ = false;
+  }
+
+  /**
+   * Set the dirty flag to indicate if the user has changed the value since
+   * the last verification.
+   * @param {boolean} dirty
+   */
+  setDirty(dirty) {
+    this.dirty_ = dirty;
+  }
+
+  /**
+   * Get the dirty flag value.
+   * @return {boolean}
+   */
+  isDirty() {
+    return this.dirty_;
+  }
+
+  /**
+   * Check if this group contains the given element.
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  containsElement(element) {
+    return this.elements_.includes(element);
+  }
+
+  /**
+   * Check if this group contains none of the given elements
+   * @param {!Array<!Element>} elements
+   * @return {boolean}
+   */
+  containsNoElementOf(elements) {
+    return !elements.some(element => this.containsElement(element));
+  }
+
+  /**
+   * Check if the group is eligible for verification.
+   */
+  shouldVerify() {
+    return this.isDirty() && this.isFilledOut();
+  }
+
+  /**
+   * Check if every required element in the group has a value,
+   * and that those values are valid.
+   */
+  isFilledOut() {
+    return this.elements_.every(element => element.checkValidity());
+  }
+
+  /**
+   * Clear the validity state of this group's elements.
+   */
+  clearErrors() {
+    this.elements_.forEach(element => element.setCustomValidity(''));
```

Comment:
The [HTML Constraints spec](https://www.w3.org/TR/html5/forms.html#constraints) specifies that [submittable elements](https://www.w3.org/TR/html5/forms.html#category-submit) are guaranteed to have `#setCustomValidity`. A publisher would have to put a non-submittable element in a verification group for this to throw, and it seems unlikely. I could check that every element in the group has a `#setCustomValidity` method in `#parseConfig`, let me know if you think that's a good idea.